services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java

```diff
@@ -0,0 +1,105 @@
+package org.keycloak.jose.jws;
+
+import org.jboss.logging.Logger;
+import org.keycloak.Token;
+import org.keycloak.TokenCategory;
+import org.keycloak.crypto.Algorithm;
+import org.keycloak.crypto.KeyUse;
+import org.keycloak.crypto.SignatureProvider;
+import org.keycloak.crypto.SignatureSignerContext;
+import org.keycloak.models.ClientModel;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.RealmModel;
+import org.keycloak.models.TokenManager;
+import org.keycloak.protocol.oidc.OIDCConfigAttributes;
+
+public class DefaultTokenManager implements TokenManager {
+
+    private static final Logger logger = Logger.getLogger(DefaultTokenManager.class);
+
+    private static String DEFAULT_ALGORITHM_NAME = Algorithm.RS256;
+
+    private final KeycloakSession session;
+
+    public DefaultTokenManager(KeycloakSession session) {
+        this.session = session;
+    }
+
+    @Override
+    public String encode(Token token) {
+        String signatureAlgorithm = signatureAlgorithm(token.getCategory());
+
+        SignatureProvider signatureProvider = session.getProvider(SignatureProvider.class, signatureAlgorithm);
+        SignatureSignerContext signer = signatureProvider.signer();
+
+        String encodedToken = new JWSBuilder().type("JWT").jsonContent(token).sign(signer);
+        return encodedToken;
+    }
+
+    @Override
+    public <T extends Token> T decode(String token, Class<T> clazz) {
+        if (token == null) {
+            return null;
+        }
+
+        try {
+            JWSInput jws = new JWSInput(token);
+
+            String signatureAlgorithm = jws.getHeader().getAlgorithm().name();
+
+            SignatureProvider signatureProvider = session.getProvider(SignatureProvider.class, signatureAlgorithm);
+            if (signatureProvider == null) {
+                return null;
+            }
+
+            String kid = jws.getHeader().getKeyId();
+            // Backwards compatibility. Old offline tokens and cookies didn't have KID in the header
+            if (kid == null) {
+                logger.debugf("KID is null in token. Using the realm active key to verify token signature.");
+                kid = session.keys().getActiveKey(session.getContext().getRealm(), KeyUse.SIG, signatureAlgorithm).getKid();
+            }
+
+            boolean valid = signatureProvider.verifier(kid).verify(jws.getEncodedSignatureInput().getBytes("UTF-8"), jws.getSignature());
+            return valid ? jws.readJsonContent(clazz) : null;
+        } catch (Exception e) {
+            logger.debug("Failed to decode token", e);
+            return null;
+        }
+    }
+
+    @Override
+    public String signatureAlgorithm(TokenCategory category) {
+        switch (category) {
+            case INTERNAL:
+                return Algorithm.HS256;
```

Comment:
Do we want this hardcoded to HS256? Wonder if someone may want stronger algorithm for refreshToken, initial access tokens and such. However we may probably add support for this in the future?

Reply:
-1 Users shouldn't be in control on how internal tokens are generated or signed. We may choose to use JWE in the future or even not use JWT at all.

Reply:
Probably yes. Just wonder that someone may claim that algorithm HS256 is not very secure and want to replace it for internal tokens with something stronger. This is probably not the case now, we can see if someone request something like that.